controller/src/main/java/io/pravega/controller/server/security/auth/AuthorizationResourceImpl.java

```diff
@@ -58,10 +60,29 @@ public String ofReaderGroupInScope(String scopeName, String readerGroupName) {
         return String.format("%s/%s:%s", ofScope(scopeName), TAG_READERGROUP, readerGroupName);
     }
 
+    @Override
+    public String ofWatermarkInScope(String scopeName, String watermarkName) {
+        Exceptions.checkNotNullOrEmpty(scopeName, "scopeName");
+        Exceptions.checkNotNullOrEmpty(watermarkName, "watermarkName");
+        return String.format("%s/%s:%s", ofScope(scopeName), TAG_WATERMARK, watermarkName);
+    }
+
     @Override
     public String ofKeyValueTableInScope(String scopeName, String keyValueTableName) {
         Exceptions.checkNotNullOrEmpty(scopeName, "scopeName");
         Exceptions.checkNotNullOrEmpty(keyValueTableName, "keyValueTableName");
         return String.format("%s/%s:%s", ofScope(scopeName), TAG_KEYVALUETABLE, keyValueTableName);
     }
+
+    @Override
+    public String ofInternalStream(String scopeName, @NonNull String streamName) {
+        // Internal stream names start with either `_RG` (reader groups) or `_MARK` (watermarks).
+        if (streamName.startsWith("_RG")) {
+            return ofReaderGroupInScope(scopeName, streamName.replace("_RG", ""));
```

Comment:
Please use some api in `NameUtils`. No code outside of that class should hardcoded these special strings.

Reply:
Done.